src/bse202/routes/user/top_up_credit.py

```diff
@@ -1,8 +1,91 @@
+card_number_regex = re.compile(r"^\d{16}$")
+expiry_date_regex = re.compile(r"^(0[1-9]|1[0-2])\/(20[2-9][0-9])$")
+cvv_regex = re.compile(r"^\d{3,4}$")
 
 
 @user_blueprint.route("/top_up_credit", methods=["GET", "POST"])
 def top_up_credit():
-    return render_template("views/top_up_credit.html")
+    if request.method == "POST":
+        # Handle form submission and validate payment details
+        card_number = request.form.get("card_number", "").strip()
+        expiry_date = request.form.get("expiry_date", "").strip()
+        cvv = request.form.get("cvv", "").strip()
+        amount = request.form.get("amount")  # Retrieve amount from form data
+
+        if not amount or not isinstance(amount, str):
+            flash("Invalid amount. Please select a valid amount.", "error")
+            return render_template("views/top_up_credit.html", amount=0)
+
+        # Convert amount to float
+        try:
+            amount = float(amount)
+        except ValueError:
+            flash("Invalid amount. Please select a valid amount.", "error")
+            return render_template("views/top_up_credit.html", amount=0)
+
+        # Ensure the inputs are strings (they should be if coming from form)
+        if not isinstance(card_number, str):
+            flash("Invalid card number format.", "error")
+            return render_template("views/top_up_credit.html", amount=amount)
+        if not isinstance(expiry_date, str):
+            flash("Invalid expiry date format.", "error")
+            return render_template("views/top_up_credit.html", amount=amount)
+        if not isinstance(cvv, str):
+            flash("Invalid CVV format.", "error")
+            return render_template("views/top_up_credit.html", amount=amount)
+
+        # Validate card number, expiry date, and CVV
+        if not card_number_regex.match(card_number):
+            return render_template("views/top_up_credit.html", amount=amount)
+        if not expiry_date_regex.match(expiry_date):
+            flash(
+                "Invalid expiry date. Please enter a valid date in MM/YYYY format.",
+                "error",
+            )
+            return render_template("views/top_up_credit.html", amount=amount)
+        if not cvv_regex.match(cvv):
+            flash("Invalid CVV. Please enter a valid 3 or 4 digit CVV.", "error")
+            return render_template("views/top_up_credit.html", amount=amount)
+
+        # If all validations pass, top up the user's account
+        try:
+            db = get_db()
+            user_id = g.token.get("user_id")  # Retrieve user_id from g.token
+
+            if not user_id:
+                flash("User ID not found in session.", "error")
+                return render_template("views/top_up_credit.html", amount=amount)
+
+            db.execute(
+                "UPDATE users SET account_balance = account_balance + ? WHERE user_id = ?",
+                (amount, user_id),
+            )
+            db.commit()
```

Comment:
you'll need to create a cursor with `db.cursor()` then do the update with the cursor

otherwise `db.commit()` and `db.rollback()` is useless